File: TemplateMethod/problem.dart
void main(List<String> args) {
  var auditTrail = AuditTrail();
  var task = TransferMoneyTask(auditTrail: auditTrail);
  task.execute();

  var anotherTask = GenerateReportTask(auditTrail: auditTrail);
  anotherTask.execute();
}

class TransferMoneyTask {
  AuditTrail _auditTrail;

  TransferMoneyTask({required auditTrail}) : _auditTrail = auditTrail;

  void execute() {
    this._auditTrail.record();
    print("Transfer Money");
  }
}

class GenerateReportTask {
  AuditTrail _auditTrail;

  GenerateReportTask({required auditTrail}) : _auditTrail = auditTrail;

  void execute() {
    this._auditTrail.record();
    print("Generate Report");
  }
}

class AuditTrail {
  void record() {
    print("Audit");
  }
}

File: TemplateMethod/problem.py
class AuditTrail:
    def record(self) -> None:
        print("Audit")


class TransferMoneyTask:
    def __init__(self, audit_trail: AuditTrail) -> None:
        self.__audit_trail = audit_trail

    def execute(self) -> None:
        self.__audit_trail.record()
        print("Transfer Money")


class GenerateReportTask:
    def __init__(self, audit_trail: AuditTrail) -> None:
        self.__audit_trail = audit_trail

    def execute(self) -> None:
        self.__audit_trail.record()
        print("Generate Report")


def main():
    audit_trail = AuditTrail()
    task = TransferMoneyTask(audit_trail=audit_trail)
    task.execute()

    anotherTask = GenerateReportTask(audit_trail=audit_trail)
    anotherTask.execute()


main()
